### ISSUE TITLE
Child routes disappear immediately instead of playing :leave animation

### ISSUE THREAD

**Issue by LecksC** [angular/angular#15477]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Child routes disappear immediately without playing animations when the parent route changes.

**Expected behavior**
Either all child routes play their :leave animation, or child routes don't get deleted until the parent route's :leave animation is done).

**Minimal reproduction of the problem with instructions**
Here's a version of the crisis center with animations added to all components to show the issue:
https://plnkr.co/edit/QKdb7ucKDX5JX4kKiOJ1?p=preview

Animating from Heroes > Crisis Center works smoothly, however when animating from Crisis Center > Heroes, the list disappears immediately without playing the ':leave' animation.

**What is the motivation / use case for changing the behavior?**
Smoothly animating nested routing.

**Please tell us about your environment:**

Windows 10, node, asp.net core, Visual Studio (or just plunkr).

* **Angular version:** 4.0.0

* **Browser:** all
 
* **Language:** TypeScript 2.2

**Comment by nicky-lenaers** [comment]
Same issue

**Comment by GProst** [comment]
Same here on Angular 4.2.2

`animateChild()` doesn't help:
In parent router animation:
 
    transition(':leave', [
      query('@*', animateChild()),
      animate('/*some easing*/', style({/*some styles*/}))
    ])

This causes error `query("@*") returned zero elements.`. So child route is removed immediately, parent component can't see it.

**Comment by GProst** [comment]
OK, I think I found the cause.

When I am trying to catch nested route from outside the root `<router-outlet>` everything is ok (as described [here](https://www.yearofmoo.com/2017/06/new-wave-of-animation-features.html#8-1-getting-router-outlet-to-animate)). I am able to set nested route animation or call its own animation via `animateChild()`.

But if I want to do the same thing from inside of the root `<router-outlet>` (bind animation to the route host component via `@HostBinding()`) the above doesn't work. Nested route disappears immediately, animation `query` can't catch it. I think the main reason is that you can't catch `outlet.activatedRouteData` with the next route path from inside of `<router-outlet>`.

Say, if I will `console.log(outlet.activatedRouteData.animation)` (see the [article](https://www.yearofmoo.com/2017/06/new-wave-of-animation-features.html#8-1-getting-router-outlet-to-animate)) from outside of the root `<router-outlet>` I will get data for all routes (for example, `postsPage`, `aboutPage`, `helpPage` etc). But if I would do the same from inside of the root `<router-outlet>` I will get only data to corresponding route (for example `postsPage`), and if I'll go to `/about` page I won't catch this event and there will be no `console.log` and no route animation.

This is inconvenient, because we are forced to make difficult animation queries from outside of the root `<router-outlet>` down to all nested routes (for each `:enter` and `:leave` components). It can be very difficult task to set queries even if you have 1 nested route (I've just experienced it in my own not very difficult project - a blog with just 2 outer routes and 1 nested route). But what if there will be more nested levels?

**Comment by daleyjem** [comment]
I'm having this issue myself.

In my experience, a `<router-outlet>` nested inside of another `<router-outlet>` is cleared of its contents during a transition, while any sibling of the nested `<router-outlet>` remains.

**Comment by matsko** [comment]
This should be fixed once #19455 lands

**Comment by ChristianSchroedel** [comment]
Still got this issue in Angular 5.1.3.

**Comment by ChristianSchroedel** [comment]
This issue is still present in 5.2.2.

Created a stackblitz to reproduce the issue with a simple use-case:
https://stackblitz.com/edit/angular-router-animations

When the transition is triggered, the content of the outer `router-outlet` is animated properly, but the content of the inner `router-outlet` disappears immediatly.

As far as I know there is already a PR with changes in the pipeline, just asking, when can we assume this to be fixed in a release? Since it was already working properly our PM just sees it as a regression we have to fix somehow ...

If you need a better sample to reproduce or more information, feel free to ask me!

**Comment by y4nnL** [comment]
The only way I've found to prevent the inner router-outlet content from being removed is to trigger a real but not visible animation at the outer outlet level, like so

```JavaScript
// a "slide to left" animation between 2 outer routes
transition('routeA => routeB', [
  group([
    query('.outer-router-outlet > :leave', [
      style({ zIndex: 2, left: 0, opacity: 1 }),
      animate('200ms ease-in', style({ left: 200, opacity: 0 }))
    ], { optional: true }),
    query('.outer-router-outlet > :enter', [
      style({ zIndex: 1, left: -200, opacity: 0 }),
      animate('200ms 100ms ease-out', style({ left: 0, opacity: 1 }))
    ], { optional: true }),
    // force the inner content to be animated
    // with a real but not visible animation
    // the inner content is not removed until the animation ends
    // we use the same 200ms
    query('.outer-router-outlet .inner-router-outlet router-outlet ~ *', [
      style({ opacity: .99999 }),
      animate(200, style({ opacity: 1 }))
    ], { optional: true })
  ])
])
```

Hope this trick helps until the PR is merged

**Comment by ChristianSchroedel** [comment]
Thank you!
Your workaround really helps me until this issue fixed :smiley:

**Comment by hmoog** [comment]
additional hint towards @y4nnL idea:

it's actually enough to just trigger an empty animation that lasts 1 millisecond on the child route for it to not disappear.

the animation var holds my animation metadata:

transition('A => B', group([
    query(':self', animation),
    query('router-outlet ~ *', [style({}), animate(1, style({}))], { optional: true })
]))

**Comment by vladimircordas** [comment]
Thanks @hmoog , your solution fixed the disappearing of the child routes, but I got a new issue: "Failed to execute 'animate' on 'Element': Partial keyframes are not supported." which happens when I switch routes while animation is not yet fully completed.
Angular 5.2.10.

**Comment by serkon** [comment]
I have got the same issue

Angular CLI: 1.7.4
Node: 9.11.1
OS: darwin x64
Angular: 5.2.10
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

@angular/cli: 1.7.4
@angular-devkit/build-optimizer: 0.3.2
@angular-devkit/core: 0.3.2
@angular-devkit/schematics: 0.3.2
@ngtools/json-schema: 1.2.0
@ngtools/webpack: 1.10.2
@schematics/angular: 0.3.2
@schematics/package-update: 0.3.2
typescript: 2.5.3
webpack: 3.11.0

**Comment by Simon-Co** [comment]
The issue is still present.  Thanks for the solutions Hans and y4nnL.

**Comment by rienheuver** [comment]
I still have this issue as well

**Comment by artuska** [comment]
Have the same issue. All those hacks above just does not work.

**Comment by JustDoItSascha** [comment]
Thanks @y4nnL , your workaround worked for me like a charm. I'm now using:

```
query(':scope > :leave router-outlet ~ *', [
      style({ opacity: .99999 }),
      animate('300ms ease-out', style({ opacity: 1 }))
    ], { optional: true })
```

**Comment by darkyelox** [comment]
No one of those hacks worked for me.

**Comment by Simon-Co** [comment]
> No one of those hacks worked for me.

Have you upped the z-index on the elements being removed to higher than the incoming elements?

**Comment by darkyelox** [comment]
@Simon-Co i don't think is that the problem, my animation is a slide from left <-> right and the component is removed when i do the navigation, then the animation start but don't show anything in the :leave process becase of the current element is removed and replaced by the new upon navigation, i need to find some way to no remove the current view while the animation is running but i don't know how

**Comment by JustDoItSascha** [comment]
The workaround from @y4nnL works, because any element which is animating can't be removed while animating. So if this workaround is not working for you, I think you have to work on the selector. I had to change it to `:scope > :leave router-outlet ~ *`, but maybe your elements are different and you need another selector.

For me the part `router-outlet ~ *` was absolutely crucial, because not the router-outlet itself is animated, but every element after router-outlet (because the components are *not rendered inside* the router-outlet.

**Comment by JustDoItSascha** [comment]
And within Step1 is another (secondary) router outlet you want to stop from
disappearing? I just want to clarify whether your problem is the same as
mine before I say something wrong ;-)

darkyelox <notifications@github.com> schrieb am Di., 15. Okt. 2019, 17:35:

**Comment by darkyelox** [comment]
@JustDoItSascha in my case this is an animation intended for slide out in :leave and slide in in :enter, lets say we do a transition from Step1 to Step2 (using internal router-outlet containing Step1 and Step2), Step1 :leave slide out animation is never shown because Step1 disappears as soon as we change the router path, but the slide in animation in Slide2 runs perfect, that is my problem.

**Comment by JustDoItSascha** [comment]
@darkyelox Sorry for the late answer. I was on holiday.

I think, that your problem is something different than this topic is about. But maybe I can help you anyway, but better you write me at something like Twitter @saschawolff1, so we don't pollute this topic ;-)

**Comment by brandon-sigao** [comment]
Still happening on 8.3.  I thought I was going crazy because its literally in the animation docs that you can do this, but the page just disappears instead of animating.  

For clarity, all I'm trying to do is run animations on a page before it leaves like so:  Click a "next" button -> have current page animations run -> transition to new page -> have new page animations run.  That's the entire point of "animateChild()" right?

**Comment by mateofacu** [comment]
Still happening on 9.1

**Comment by bcerquiare** [comment]
Still on 10.0.3

**Comment by arctouch-viniciusbarbosa** [comment]
Still happening! :/

**Comment by Abasz** [comment]
> additional hint towards @y4nnL idea:
> 
> it's actually enough to just trigger an empty animation that lasts 1 millisecond on the child route for it to not disappear.
> 
> the animation var holds my animation metadata:
> 
> ```
> let animation = [...];
> 
> transition('A => B', group([
>     query(':self', animation),
>     query('router-outlet ~ *', [style({}), animate(1, style({}))], { optional: true })
> ]))
> ```

The issue still persists, the above workaround fixes this temporarily.

**Comment by cchulo** [comment]
> > additional hint towards @y4nnL idea:
> > it's actually enough to just trigger an empty animation that lasts 1 millisecond on the child route for it to not disappear.
> > the animation var holds my animation metadata:
> > ```
> > let animation = [...];
> > 
> > transition('A => B', group([
> >     query(':self', animation),
> >     query('router-outlet ~ *', [style({}), animate(1, style({}))], { optional: true })
> > ]))
> > ```
> 
> The issue still persists in version 11+, the above workaround works temporarily.

I'm also working on Angular 11, think you can post a stackblitz to see how you got yours working?

**Comment by cchulo** [comment]
i figured something out. Note that I'm on Angular 11.

Here is my code:
```
let animation = [
    style({position: 'relative'}), // per the docs, the host view needs to be relative
    query(':enter, :leave', [ // child views need to be absolute
        style({
            position: 'absolute',
            top: 0,
            right: 0,
            width: '100%'
        })
    ], { optional: true }),
    query(':enter', [ // the new view that is entering should be invisible
        style({ opacity: 0 })
    ]),
    query(':leave', query('@*', animateChild(), { optional: true }), { optional: true }), // play leave animations for view that is being removed
    group([
        query(':leave', useAnimation(zoomOut), { optional: true }), // fade out the view that is leaving
        query(':enter', useAnimation(bounceIn)) // fade in the new view
    ]),
    query(':enter', query('@*', animateChild(), { optional: true }),) // play animations for the new view
];

export const fadeAnimation = 
trigger('routeAnimation', [
    transition('Home <=> *', animation),
    transition('Blog <=> *', animation),
    transition('Projects <=> *', animation),
    transition('About <=> *', animation),
    transition('Contact <=> *', animation),
]);
```

I have 6 pages for my web app: Home, Blog, Projects, About and Contact.

The goal is to unload a page using a zoom out animation, and introduce a new view by bouncing it in. In case you are wondering, the zoomOut and bounceIn animations are from the ng-animate package found [here](https://github.com/jiayihu/ng-animate). I also wanted, before the old view would zoom out, animate the old components' individual leave animations, and when the new view bounced in play the component's enter animations. This wasn't happening   though, and is what brought me to this thread in the first place.

The zoomOut/bounceIn animations were playing, but not the animateChild() calls. I tried the fixes listed on this thread, but what finally what worked is the code snippet above. This was all thanks to a blog post found [here](https://indepth.dev/posts/1285/in-depth-guide-into-animations-in-angular) that ultimately gave me a better understanding of how the animation system works.

So on the view that is leaving, I do a wild card query using @* and call animate child on that. The {optional: true} is required because on initial page load there is no such leave view. This is not necessary for :enter, but i left it in there anyway it doesn't really matter.

Now whenever I switch between pages in any order, the leave animations in the view's children play correctly before the zoomOut animation plays, and then the new view is bouncedIn and the new views' children components play their :enter animations correctly.

here is how you use it:

```
<div class="content" [@routeAnimation]="prepareRoute(outlet)">
  <router-outlet #outlet="outlet"></router-outlet>
</div>
```

The prepareRoute is identical to the one in Angular's tutorial, nothing special there. Hope this helps someone.

**Comment by dolanmiu** [comment]
Still a problem in Angular 12!

@cchulo I have created a StackBlitz to demonstrate the workaround @y4nnL and @hmoog suggested:

### With Fix
https://stackblitz.com/edit/angular-router-animations-lry4sx?file=app%2Fanimations.ts

### Without Fix
https://stackblitz.com/edit/angular-router-animations

**Comment by SmailHammour** [comment]
> https://stackblitz.com/edit/angular-router-animations-lry4sx?file=app%2Fanimations.ts

Omg thank you! I've been looking for a fix for ages and this is the only one that works!

**Comment by SmailHammour** [comment]
> additional hint towards @y4nnL idea:
> 
> it's actually enough to just trigger an empty animation that lasts 1 millisecond on the child route for it to not disappear.
> 
> the animation var holds my animation metadata:
> 
> ```
> let animation = [...];
> 
> transition('A => B', group([
>     query(':self', animation),
>     query('router-outlet ~ *', [style({}), animate(1, style({}))], { optional: true })
> ]))
> ```

Omg thank you! I've been looking for a fix for ages and this is the only one that works!

**Comment by thePunderWoman** [comment]
This issue was closed as obsolete due to the introduction of the new `animate.enter` and `animate.leave` APIs. The `@angular/animations` package is now deprecated and will be removed in Angular v23. We hope you give the new APIs a try.
